File: 2020/2020W17/preppindataw17.R
library(dplyr)
library(readxl)
library(tidyr)
library(stringr)
library(splitstackshape)

xlsx <- "E:/PD 2020W17 Shows and Devices.xlsx"
show_list <- read_excel(xlsx, sheet = "Netflix Shows") %>%
  distinct() %>%
  mutate('Show' = str_trim (toupper(str_extract(`TV Shows/Movies`,'([^\\(]+)'))))

dev_list <- read_excel(xlsx, sheet = "Devices") %>%
  distinct() %>%
  mutate('Using' = tolower(`Device`))

survey <- read_excel('E:/PD 2020W17 Survey.xlsx', sheet = 1) %>%
  select(-'Timestamp') %>%
  distinct() 

device <- survey %>%
  select('How have you been watching Netflix? (Phone, TV, etc.)') %>%
  rename('Using' = !!names(.[1])) %>%
  cSplit(., 'Using', ',|&') %>%
  pivot_longer(.,cols=matches('^Using'), names_to='ToDrop', values_to='Using', values_drop_na = TRUE) %>%
  mutate('Using' = tolower(`Using`)) %>%
  filter(`Using`!='etc.') %>%
  merge(.,dev_list,by='Using',all.x=TRUE) %>%
  mutate_at(vars('Device'), ~replace_na(., 'Other')) %>%
  group_by(`Device`) %>%
  summarise('Count' = n())

show_in_Q <- colnames(survey)[grepl('.*rate.*\\w\\?', colnames(survey))] %>%
  str_extract(.,'(?<=rate\\s)(.*)(?=\\?)') %>%
  toupper(.)

show <- rbind(survey %>%
  select(c('How would you rate "Other"?','What have you been binging during lockdown?')) %>%
  rename('Rate' = !!names(.[1]), 'lockdown' = !!names(.[2])) %>%
  cSplit(., 'lockdown', ',') %>%
  pivot_longer(.,cols=matches('^lockdown'), names_to='ToDrop', values_to='Show', values_drop_na = TRUE) %>%
  mutate('Show'= toupper(`Show`)) %>%
  filter(is.na(match(`Show`,show_in_Q))) %>%
  merge(.,show_list,by='Show') %>%
  select(c('Show','Rate')),
  survey %>%
  select(matches("lockdown") | matches("rate [^\"]")) %>%
  rename('lockdown' = !!names(.[1])) %>%
  pivot_longer(.,cols=matches('rate'), names_to='Question', values_to='Rate', values_drop_na = TRUE) %>%
  mutate('Show' =  toupper(str_extract(`Question`,'(?<=rate\\s)(.*)(?=\\?)'))) %>%
  filter(str_detect(toupper(`lockdown`),`Show`)) %>%
  select(c('Show','Rate'))) %>%
  group_by(`Show`) %>%
  summarise('Rate' = mean(`Rate`)) %>%
  mutate('Rank' = dense_rank(desc(`Rate`)))

View(device)
View(show)

File: 2020/2020W17/PreppindataW17.py
import pandas as pd
import numpy as np
import re

xls = pd.ExcelFile("E:/PD 2020W17 Shows and Devices.xlsx")
dev_list = pd.read_excel(xls,'Devices')
dev_list['Device_lower'] = dev_list['Device'].str.lower()
show_list = pd.read_excel(xls,'Netflix Shows').drop_duplicates()
show_list['Show'] = show_list['TV Shows/Movies'].apply(lambda x: re.search("([^\(]+)",str(x)).group(1).strip().upper())

survey = pd.read_excel(pd.ExcelFile("E:/PD 2020W17 Survey.xlsx"),sheet_name=0)
survey = survey.drop('Timestamp', axis=1).drop_duplicates()

device = pd.DataFrame([map(str.strip, map(str.lower,x)) for x in survey['How have you been watching Netflix? (Phone, TV, etc.)']
                     .str.split(',|&').values.tolist()])
device = device.melt(value_name='Device_lower', var_name='To_drop').replace({'Device_lower': {'etc.': np.nan}}).drop('To_drop', axis=1).dropna()
device = pd.merge(device, dev_list, on='Device_lower', how='left',indicator='Join')
device['Device'] = device.apply(lambda x:x['Device'] if x['Join']=='both' else 'Other', axis=1)
device = device.groupby('Device', as_index=False).agg({'Join':'count'}).rename(columns={'Join':'Count'})


show_in_Q = [re.search('rate ([^\?]+)(?=\?)',c).group(1).upper() for c in survey.columns if re.search('rate ([\w\s\-]+)(?=\?)',c)]
show = pd.concat([survey[['How would you rate "Other"?']].rename(columns={'How would you rate "Other"?':'Rate'}), 
                  pd.DataFrame([map(str.strip, map(str.upper,x)) for x in survey['What have you been binging during lockdown?']
                                .str.split(',').values.tolist()])], axis=1, sort=False)
show = show.melt(id_vars='Rate', value_name='Show', var_name='To_drop').drop('To_drop', axis=1).dropna()
show = pd.merge(show, show_list, on='Show')
show['Is Other'] = show['Show'].apply(lambda x: str(x) not in show_in_Q)
show = show[show['Is Other']][['Show','Rate']].copy()

showQ = survey[[c for c in survey.columns if re.search('rate ([\w\s\-]+)(?=\?)',c) or re.search('(lockdown)',c)]].rename(
    columns={'What have you been binging during lockdown?':'lockdown'})
showQ = showQ.melt(id_vars='lockdown', value_name='Rate', var_name='Show').dropna()
showQ['Show'] = showQ['Show'].apply(lambda x: re.search('rate ([^\?]+)(?=\?)',x).group(1).upper())
showQ['Is Lockdown'] = showQ.apply(lambda x: True if re.search(f'({x["Show"]})',x['lockdown'],re.I) else False, axis=1)

show = show.append(showQ[showQ['Is Lockdown']][['Show','Rate']].copy())
show = show.groupby('Show', as_index=False).agg({'Rate':'mean'})
show['Rank'] = show['Rate'].rank(method='dense', ascending=False).astype(int)
